src/Hazelcast.Net/Serialization/Compact/ReflectionSerializer.cs

```diff
@@ -399,80 +399,153 @@ private static PropertyInfo[] GetProperties(Type objectType)
+                    continue;
+
+                // read the field value and set the property value accordingly
+                property.SetValue(obj, GetPropertyValue(reader, property.PropertyType, fieldName));
             }
-            catch (Exception e)
+
+            return obj;
+        }
+
+        // constructs an object of a specified type, using the empty constructor if possible,
+        // otherwise using the constructor with most (and all) parameters matching field names
+        // (case-sensitive) - and then ctorFields contains the fields that have been consumed
+        private static object ConstructObject(Type type, CompactReader reader, out HashSet<string>? ctorFields)
+        {
+            // value types can always be created by the activator regardless of their constructors
+            // and, they don't implicitly implement an empty constructor - so we *want* to use the activator
+            if (type.IsValueType)
             {
-                throw new SerializationException($"Failed to create an instance of type {typeOfObj}.", e);
+                ctorFields = null;
+                try
+                {
+                    var obj = Activator.CreateInstance(type);
+                    if (obj == null)
+                        throw new SerializationException($"Failed to create an instance of type {type}.");
+                    return obj;
+                }
+                catch (Exception e)
+                {
+                    throw new SerializationException($"Failed to create an instance of type {type}.", e);
+                }
             }
 
-            if (obj == null)
-                throw new SerializationException($"Failed to create an instance of type {typeOfObj}.");
+            var ctors = type.GetConstructors();
 
-            // TODO: consider emitting the property setters
+            // look for the empty constructor and use it
+            var emptyCtor = ctors.FirstOrDefault(ctor => ctor.GetParameters().Length == 0);
+            if (emptyCtor != null)
+            {
+                try
+                {
+                    ctorFields = null;
+                    return emptyCtor.Invoke(Array.Empty<object>());
+                }
+                catch (Exception e)
+                {
+                    throw new SerializationException($"Failed to create an instance of type {type}.", e);
+                }
+            }
+
+            // look for the constructor with most (and all) parameters matching field names (case-sensitive)
+            // ReSharper disable once SimplifyLinqExpressionUseMinByAndMaxBy
+            var ctor = ctors
+                .Where(ctor => ctor.GetParameters().All(x => x.Name != null && reader.ValidateFieldName(x.Name)))
+                .OrderBy(x => x.GetParameters())
+                .LastOrDefault();
+
+            if (ctor == null)
+                throw new SerializationException($"Failed to create an instance of type {type}.");
```

Comment:
Since exception is obvious, message could be more explicit. Like "Failed to create an instance of type {type} because proper constructor not found." WDYT?

Reply:
make sense, updating.